mdns: update if-watch to 3.0.0

jxs opened the pull request:
and rename TokioMdns to Mdns living in its own module, and move Mdns to async_io::Mdns.

# Description

updates to `if-watch` `3.0`. This was simplest I found to not touch the signature of `GenMdns::new` wdyt @thomaseizinger? 
@mxinden turns out `IfWatcher::new` returns `Result`  so implementing `Default` for it wouldn't solve the problem, so I had to create a new `TryDefault` trait to be able to abstract over it.

We could also try put  `GenMdns` methods and trait impls on a macro and generate it for `tokio` and `async_io` code would be probably more complicated, but the doc would be cleaner and `GenMdns` would no longer be required.

CC @rkuhn 
## Links to any relevant issues
superseeds #3095
<!-- Reference any related issues.-->


## Open Questions

<!-- Unresolved questions, if any. -->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

<!-- The below text will appear as the commit message body once we squash-merge the PR. -->
## Commit message body

jxs committed:
mdns: update if-watch to 3.0,

and rename TokioMdns to Mdns living in its own module,
and move Mdns to async_io::Mdns.

rkuhn reviewed (commented):
Looks good to me for mDNS! Shall I do the TCP part or do you want to take it?

jxs committed:
mdns: update examples to the new Mdns.

thomaseizinger commented on protocols/mdns/CHANGELOG.md:
```diff
@@ -4,6 +4,9 @@
 
 - Update to `libp2p-swarm` `v0.41.0`.
 
+- Update to `if-watch` `0.3.0` and both rename `TokioMdns` to `Mdns` living in `tokio::Mdns`,
```
```suggestion
- Update to `if-watch` `3.0.0` and both rename `TokioMdns` to `Mdns` living in `tokio::Mdns`,
```

jxs replied:
Thanks Thomas, updated.

thomaseizinger commented on protocols/mdns/src/lib.rs:
```diff
@@ -42,10 +42,10 @@ mod behaviour;
 
 #[cfg(feature = "async-io")]
-pub use crate::behaviour::Mdns;
+pub use crate::behaviour::async_io;
 
 #[cfg(feature = "tokio")]
-pub use crate::behaviour::TokioMdns;
+pub use crate::behaviour::tokio;
```
If we already do this, should we just go all the way and rename `Mdns` to `Behaviour`, following #2217?

jxs replied:
yeah, agree! :)

jxs replied:
done :)

thomaseizinger commented on protocols/mdns/src/behaviour.rs:
```diff
@@ -36,29 +36,53 @@ use smallvec::SmallVec;
+        fn try_default() -> Result<Self, std::io::Error> {
+            Self::new()
+        }
+    }
```
An alternative way of doing this which does not require a trait would be to say:

```rust
impl GenMdns<TokioUdpSocket, TokioTimer, IfWatcher> {
    /// Builds a new `Mdns` behaviour.
    pub fn new(config: MdnsConfig) -> io::Result<Self> {
        Ok(Self {
            config,
            if_watch: IfWatcher::new(),
            iface_states: Default::default(),
            discovered_nodes: Default::default(),
            closest_expiration: Default::default(),
        })
	}
}
```

thomaseizinger replied:
Untested but I think this should work!

jxs replied:
Yeah, that works if we remove `GenMdns::new` otherwise compiler complains regarding duplication. The disadvantage I see with that is that then methods (and its docs) are scattered between `{tokio, async_io}::Mdns` and `GenMdns`, would you prefer that approach? Tbh ideally I think it would be great if we could deprecate/hide `GenMdns` and just present `{tokio, async_io}::Mdns` to the end user.

thomaseizinger replied:
We had a similar discussion in `libp2p-tcp`: https://github.com/libp2p/rust-libp2p/pull/2961#pullrequestreview-1129815307

From what I remember, the functions don't show up for the type alias at all so probably best to not do this at all.

The need for another trait is a bit annoying though.

thomaseizinger reviewed (commented):
Thanks for tackling this @jxs!

jxs committed:
review: fix typo in if-watch version on CHANGELOG

thomaseizinger commented on protocols/mdns/src/behaviour.rs:
```diff
@@ -36,29 +36,53 @@ use smallvec::SmallVec;
 
     /// Iface watcher.
-    if_watch: IfWatcher,
+    if_watch: W,
```
Can we do `T::IfWatcher` instead, similar to what @rkuhn did for TCP?

jxs replied:
Yeah, great suggestion! Ptal if that's what you were expecting Thomas.

thomaseizinger commented on protocols/mdns/CHANGELOG.md:
```diff
@@ -4,6 +4,9 @@
 - Update to `libp2p-swarm` `v0.41.0`.
 
+- Update to `if-watch` `3.0.0` and both rename `TokioMdns` to `Mdns` living in `tokio::Mdns`,
+and move `Mdns` to `async_io::Mdns`. see [PR 3096]
```
```suggestion
and move `Mdns` to `async_io::Mdns`. See [PR 3096].
```

jxs replied:
Thanks Thomas, updated!

mxinden reviewed (commented):
Looks good to me minus the outstanding comments.

jxs committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into update-if-watch

jxs committed:
review: use Associated Type for the aux trait,

 instead of a Generic parameter

jxs committed:
review: remove  prefixes from types

thomaseizinger commented on protocols/mdns/CHANGELOG.md:
```diff
@@ -4,6 +4,16 @@
 - Update to `libp2p-swarm` `v0.41.0`.
 
+- Update to `if-watch` `3.0.0` and both rename `TokioMdns` to `Behaviour` living in `tokio::Behaviour`,
+and move and rename `Mdns` to `async_io::Behaviour`. See [PR 3096]
```
```suggestion
and move and rename `Mdns` to `async_io::Behaviour`. See [PR 3096].
```

jxs replied:
Thanks, addressed.

thomaseizinger commented on protocols/mdns/src/behaviour.rs:
```diff
@@ -36,29 +36,63 @@ use smallvec::SmallVec;
 /// them to the topology.
 #[derive(Debug)]
-pub struct GenMdns<S, T> {
+pub struct Behaviour<S, T>
```
Would it make sense to combine all type parameters here to a single one, i.e. `P` and have a `Provider` trait?

```rust
trait Provider {
	type Socket;
	type Timer;
	type IfWatcher;
}
```

jxs replied:
indeed it does! Thanks Thomas. :+1:  
We now have more than one `Provider` trait among the protocols (`tcp` and `mdns`) but as they are namedspaced by the crates I don't think it's a problem, it can be considered akin to the `tokio` and `async_io` namespaces.

thomaseizinger commented on protocols/mdns/src/lib.rs:
```diff
@@ -38,14 +38,26 @@
 mod behaviour;
-pub use crate::behaviour::{GenMdns, MdnsEvent};
+pub use crate::behaviour::{Behaviour, Event};
 
 #[cfg(feature = "async-io")]
-pub use crate::behaviour::Mdns;
+pub use crate::behaviour::async_io;
 
 #[cfg(feature = "tokio")]
-pub use crate::behaviour::TokioMdns;
```
If you'd add type aliases for this too then users would only be presented with deprecation warnings instead of compile errors.

jxs replied:
yup, makes sense. When putting the types was thinking that since this is a breaking change putting the types would be redundant, but if we also add for `Mdns` then it doesn't become a breaking change. Thanks Thomas :)

thomaseizinger reviewed (commented):
Thanks @jxs! A few comments.

jxs committed:
review: add missing dot on CHANGELOG.

jxs committed:
review: add type alias for previous Mdns type.

jxs committed:
review: turn WatcherProvider into Provider,

and set the AsyncSocket and AsyncTimer as associated types, so that the
generic Behaviour only takes one parameter P which provides all these
types.

jxs committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into update-if-watch

thomaseizinger commented on protocols/mdns/CHANGELOG.md:
```diff
@@ -4,8 +4,17 @@
+
+- Remove the remaning `Mdns` prefixes from types as per [discussion 2174].
+  I.e the `Mdns` prefix has been removed from various types like `MdnsEvent`.
+  Users should prefer importing the identify protocol as a module (`use libp2p::mdns;`),
```
```suggestion
  Users should prefer importing the mdns protocol as a module (`use libp2p::mdns;`),
```

jxs replied:
omg.. :man_facepalming: :see_no_evil: Thanks Thomas fixed it!

thomaseizinger commented on protocols/mdns/CHANGELOG.md:
```diff
@@ -22,6 +31,7 @@
 
 - Update to `if-watch` `v2.0.0` and thus the `async` method `Mdns::new` and `TokioMdns::new` becomes synchronous. See [PR 2978].
 
+[PR 3096]: https://github.com/libp2p/rust-libp2p/pull/3096
```
This link is in the wrong section, should be under `# 0.42.0`.

jxs replied:
Right, thanks Thomas, moved it.

thomaseizinger commented on protocols/mdns/src/behaviour.rs:
```diff
@@ -36,32 +36,80 @@ use smallvec::SmallVec;
 
-#[cfg(feature = "async-io")]
-use crate::behaviour::{socket::asio::AsyncUdpSocket, timer::asio::AsyncTimer};
+/// The interface for non-blocking TCP I/O providers.
```
```suggestion
/// An abstraction to allow for compatibility with various async runtimes.
```

jxs replied:
updated, thanks.

thomaseizinger commented on protocols/mdns/src/behaviour.rs:
```diff
@@ -72,19 +120,19 @@ pub struct GenMdns<S, T> {
 where
-    T: Builder,
+    P: Provider,
+    P::Timer: Builder,
```
I think these could be specified on the trait itself, then you wouldn't have to repeat them here.

jxs replied:
Yup, you are right Thomas, thanks. Removed the redundancy and moved both here and in the `impl` of the trait to the trait definition ptal.

thomaseizinger reviewed (approved):
Nice!

jxs committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into update-if-watch

jxs committed:
review: address Thomas suggestions,

- Remove trait bounds where they are redundant.
- Improve documentation of `Provider` trait.

jxs committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into update-if-watch

jxs committed:
Merge branch 'master' into update-if-watch

thomaseizinger commented on protocols/mdns/CHANGELOG.md:
```diff
@@ -4,11 +4,23 @@
+  and refer to its types via `mdns::`. For example: `mdns::Behaviour` or `mdns::Event`.
+
+- Use `trust-dns-proto` to parse DNS messages. See [PR 3102].
+
```
```suggestion
```

Duplicate entry :)

thomaseizinger reviewed (approved):
Very nice & clean!

Good to go from my end :)

thomaseizinger committed:
Update protocols/mdns/CHANGELOG.md
